pkg/querier/worker_service_test.go

```diff
@@ -0,0 +1,218 @@
+package querier
+
+import (
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	querier_worker "github.com/cortexproject/cortex/pkg/querier/worker"
+	"github.com/gorilla/mux"
+	"github.com/grafana/dskit/services"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/weaveworks/common/middleware"
+)
+
+func Test_InitQuerierService(t *testing.T) {
+	requestedAuthenticated := false
+	noopWrapper := middleware.Func(func(next http.Handler) http.Handler {
+		requestedAuthenticated = true
+		return next
+	})
+
+	var mockQueryHandlers = map[string]http.Handler{
+		"/loki/api/v1/query": http.HandlerFunc(func(res http.ResponseWriter, req *http.Request) {
+			_, err := res.Write([]byte("test handler"))
+			require.NoError(t, err)
+		}),
+	}
+
+	testContext := func(config WorkerServiceConfig) (*mux.Router, services.Service) {
+		requestedAuthenticated = false
+		externalRouter := mux.NewRouter()
+		querierWorkerService, err := InitWorkerService(config, mockQueryHandlers, externalRouter, http.HandlerFunc(externalRouter.ServeHTTP), noopWrapper)
+		require.NoError(t, err)
+
+		return externalRouter, querierWorkerService
+	}
+
+	t.Run("when querier is configured to run standalone, without a query frontend", func(t *testing.T) {
+		t.Run("register the internal query handlers externally", func(t *testing.T) {
+			config := WorkerServiceConfig{
+				QueryFrontendEnabled:  false,
+				QuerySchedulerEnabled: false,
+				AllEnabled:            false,
+				QuerierWorkerConfig:   &querier_worker.Config{},
+			}
+
+			externalRouter, _ := testContext(config)
+
+			recorder := httptest.NewRecorder()
+			request := httptest.NewRequest("GET", "/loki/api/v1/query", nil)
+			externalRouter.ServeHTTP(recorder, request)
+			assert.Equal(t, 200, recorder.Code)
+			assert.Equal(t, "test handler", recorder.Body.String())
+		})
+
+		t.Run("wrap external handler with auth middleware", func(t *testing.T) {
```

Comment:
This took me a few minutes to understand, the test name implied something was being done but after looking around I was able to see that every test actually has the auth middleware.

The other part of this I'm a little worried about is the use of a global variable, this feels a little brittle to me over time.

What if instead the `testContext` function took a config and a middleware.Interface which you could then provide a specific version of for this test which does what the current noop middleware does but it would be scoped within this test.

All the other tests could then just call `testContext(config, nil)` or you could make a separate `testContextWithMiddleware(config, middleware.Interface)` and have testContext call that with a nil middleware to save refactoring all your tests

WDYT?

Reply:
I like it! IMO a test should always be easier to understand than the code it's testing, so I love simplifications and clarifications.